src/device/sendrecv.h

```diff
@@ -14,12 +14,14 @@
 #endif
 
 template<typename T, typename RedOp>
-struct RunWork<ncclFuncSendRecv, T, RedOp, NCCL_ALGO_RING, NCCL_PROTO_SIMPLE> {
+struct RunWorkBatch<ncclFuncSendRecv, T, RedOp, NCCL_ALGO_RING, NCCL_PROTO_SIMPLE> {
+  static_assert(sizeof(T)==1, "SendRecv only works on single byte types T.");
+
   template<typename Proto>
-  __device__ void runSend(const int tid, const int nthreads, const uint8_t group, struct ncclWorkElemP2p* args) {
+  __device__ void runSend(int tid, int tn, int group, struct ncclDevWorkP2p* work) {
 #if defined(ENABLE_NPKIT)
     bool isNpKitThread = (tid == 0);
-    int npKitCtxIdx = blockIdx.x * NCCL_MAX_WORK_ELEMENTS_P2P + group;
+    int npKitCtxIdx = blockIdx.x * 16 + (group - (1<<16));
```

Comment:
Seems we should use `NCCL_MAX_GROUPS` in `device.h`? Also, why do we need to subtract 1<<16 from group?

Reply:
Resolved in latest commit.